File: examples/api_client.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"
)

// ReactorcideClient is a simple client for the Reactorcide API
type ReactorcideClient struct {
	BaseURL string
	Token   string
	Client  *http.Client
}

// NewReactorcideClient creates a new API client
func NewReactorcideClient(baseURL, token string) *ReactorcideClient {
	return &ReactorcideClient{
		BaseURL: baseURL,
		Token:   token,
		Client: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// Job represents a job in the system
type Job struct {
	JobID       string            `json:"job_id"`
	Name        string            `json:"name"`
	Description string            `json:"description"`
	Status      string            `json:"status"`
	GitURL      string            `json:"git_url,omitempty"`
	GitRef      string            `json:"git_ref,omitempty"`
	SourceType  string            `json:"source_type"`
	JobCommand  string            `json:"job_command"`
	RunnerImage string            `json:"runner_image,omitempty"`
	JobEnvVars  map[string]string `json:"job_env_vars,omitempty"`
	CreatedAt   time.Time         `json:"created_at"`
	UpdatedAt   time.Time         `json:"updated_at"`
}

// CreateJobRequest represents the request to create a new job
type CreateJobRequest struct {
	Name        string            `json:"name"`
	Description string            `json:"description,omitempty"`
	GitURL      string            `json:"git_url,omitempty"`
	GitRef      string            `json:"git_ref,omitempty"`
	SourceType  string            `json:"source_type"`
	SourcePath  string            `json:"source_path,omitempty"`
	JobCommand  string            `json:"job_command"`
	RunnerImage string            `json:"runner_image,omitempty"`
	JobEnvVars  map[string]string `json:"job_env_vars,omitempty"`
	QueueName   string            `json:"queue_name,omitempty"`
}

// doRequest performs an HTTP request with authentication
func (c *ReactorcideClient) doRequest(method, path string, body interface{}) (*http.Response, error) {
	var bodyReader io.Reader
	if body != nil {
		jsonBody, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal request body: %w", err)
		}
		bodyReader = bytes.NewBuffer(jsonBody)
	}

	req, err := http.NewRequest(method, c.BaseURL+path, bodyReader)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+c.Token)

	return c.Client.Do(req)
}

// CheckHealth checks the API health
func (c *ReactorcideClient) CheckHealth() error {
	resp, err := c.doRequest("GET", "/api/v1/health", nil)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("health check failed: %s", body)
	}

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return err
	}

	fmt.Printf("API Health: %+v\n", result)
	return nil
}

// CreateJob creates a new job that will be submitted to Corndogs
func (c *ReactorcideClient) CreateJob(req *CreateJobRequest) (*Job, error) {
	resp, err := c.doRequest("POST", "/api/v1/jobs", req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusCreated {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("failed to create job: %s", body)
	}

	var job Job
	if err := json.NewDecoder(resp.Body).Decode(&job); err != nil {
		return nil, err
	}

	return &job, nil
}

// GetJob retrieves a job by ID
func (c *ReactorcideClient) GetJob(jobID string) (*Job, error) {
	resp, err := c.doRequest("GET", "/api/v1/jobs/"+jobID, nil)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("failed to get job: %s", body)
	}

	var job Job
	if err := json.NewDecoder(resp.Body).Decode(&job); err != nil {
		return nil, err
	}

	return &job, nil
}

// ListJobs lists all jobs
func (c *ReactorcideClient) ListJobs() ([]Job, error) {
	resp, err := c.doRequest("GET", "/api/v1/jobs", nil)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("failed to list jobs: %s", body)
	}

	var result struct {
		Jobs []Job `json:"jobs"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	return result.Jobs, nil
}

// CancelJob cancels a running job
func (c *ReactorcideClient) CancelJob(jobID string) error {
	resp, err := c.doRequest("PUT", "/api/v1/jobs/"+jobID+"/cancel", nil)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("failed to cancel job: %s", body)
	}

	return nil
}

// DeleteJob deletes a job
func (c *ReactorcideClient) DeleteJob(jobID string) error {
	resp, err := c.doRequest("DELETE", "/api/v1/jobs/"+jobID, nil)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("failed to delete job: %s", body)
	}

	return nil
}

// WaitForJob waits for a job to complete or fail
func (c *ReactorcideClient) WaitForJob(jobID string, timeout time.Duration) (*Job, error) {
	deadline := time.Now().Add(timeout)

	for time.Now().Before(deadline) {
		job, err := c.GetJob(jobID)
		if err != nil {
			return nil, err
		}

		// Check if job is in a terminal state
		switch job.Status {
		case "completed", "failed", "cancelled":
			return job, nil
		}

		// Wait before polling again
		time.Sleep(2 * time.Second)
	}

	return nil, fmt.Errorf("timeout waiting for job %s", jobID)
}

func main() {
	// Example usage
	// Note: You need to set these environment variables or update the values
	apiURL := os.Getenv("REACTORCIDE_API_URL")
	if apiURL == "" {
		apiURL = "http://localhost:8080"
	}

	apiToken := os.Getenv("REACTORCIDE_API_TOKEN")
	if apiToken == "" {
		fmt.Println("Please set REACTORCIDE_API_TOKEN environment variable")
		fmt.Println("To create a token for testing, see the test examples in coordinator_api/test/")
		os.Exit(1)
	}

	client := NewReactorcideClient(apiURL, apiToken)

	// Check API health
	fmt.Println("Checking API health...")
	if err := client.CheckHealth(); err != nil {
		fmt.Printf("Health check failed: %v\n", err)
		os.Exit(1)
	}

	// Create a simple echo job
	fmt.Println("\nCreating a job...")
	job, err := client.CreateJob(&CreateJobRequest{
		Name:        "Example Echo Job",
		Description: "A simple job that echoes a message",
		SourceType:  "git",
		GitURL:      "https://github.com/example/repo.git",
		GitRef:      "main",
		JobCommand:  "echo 'Hello from Reactorcide!'",
		RunnerImage: "alpine:latest",
		JobEnvVars: map[string]string{
			"EXAMPLE_VAR": "example_value",
		},
		QueueName: "default",
	})

	if err != nil {
		fmt.Printf("Failed to create job: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Created job: %s (status: %s)\n", job.JobID, job.Status)
	fmt.Println("Job has been submitted to Corndogs queue for processing")

	// List jobs
	fmt.Println("\nListing jobs...")
	jobs, err := client.ListJobs()
	if err != nil {
		fmt.Printf("Failed to list jobs: %v\n", err)
	} else {
		for _, j := range jobs {
			fmt.Printf("- %s: %s (status: %s)\n", j.JobID, j.Name, j.Status)
		}
	}

	// Wait for job completion (with timeout)
	fmt.Printf("\nWaiting for job %s to complete...\n", job.JobID)
	completedJob, err := client.WaitForJob(job.JobID, 5*time.Minute)
	if err != nil {
		fmt.Printf("Error waiting for job: %v\n", err)
	} else {
		fmt.Printf("Job completed with status: %s\n", completedJob.Status)
	}

	// Clean up - delete the job
	fmt.Printf("\nDeleting job %s...\n", job.JobID)
	if err := client.DeleteJob(job.JobID); err != nil {
		fmt.Printf("Failed to delete job: %v\n", err)
	} else {
		fmt.Println("Job deleted successfully")
	}
}

File: examples/api_client.py
#!/usr/bin/env python3
"""
Reactorcide API Client Example

This example demonstrates how to interact with the Reactorcide API to:
- Create jobs that are submitted to Corndogs
- Monitor job status
- List and manage jobs

Requirements:
    pip install requests
"""

import os
import sys
import time
import json
from typing import Dict, List, Optional, Any
import requests


class ReactorcideClient:
    """Simple client for the Reactorcide API"""

    def __init__(self, base_url: str, token: str):
        """
        Initialize the Reactorcide client.

        Args:
            base_url: The base URL of the Reactorcide API (e.g., http://localhost:8080)
            token: The API authentication token
        """
        self.base_url = base_url.rstrip('/')
        self.token = token
        self.session = requests.Session()
        self.session.headers.update({
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json'
        })

    def check_health(self) -> Dict[str, Any]:
        """Check the API health status."""
        response = self.session.get(f'{self.base_url}/api/v1/health')
        response.raise_for_status()
        return response.json()

    def create_job(self,
                   name: str,
                   job_command: str,
                   source_type: str = 'git',
                   git_url: Optional[str] = None,
                   git_ref: Optional[str] = None,
                   source_path: Optional[str] = None,
                   description: Optional[str] = None,
                   runner_image: str = 'alpine:latest',
                   job_env_vars: Optional[Dict[str, str]] = None,
                   queue_name: str = 'default',
                   timeout_seconds: Optional[int] = None,
                   priority: Optional[int] = None) -> Dict[str, Any]:
        """
        Create a new job that will be submitted to Corndogs.

        Args:
            name: The name of the job
            job_command: The command to execute
            source_type: Type of source ('git' or 'copy')
            git_url: Git repository URL (for git source type)
            git_ref: Git reference (branch/tag/commit)
            source_path: Path to source (for copy source type)
            description: Job description
            runner_image: Docker image to use for execution
            job_env_vars: Environment variables for the job
            queue_name: Corndogs queue name
            timeout_seconds: Job timeout in seconds
            priority: Job priority (higher = more important)

        Returns:
            The created job object
        """
        payload = {
            'name': name,
            'job_command': job_command,
            'source_type': source_type,
            'runner_image': runner_image,
            'queue_name': queue_name
        }

        if description:
            payload['description'] = description
        if git_url:
            payload['git_url'] = git_url
        if git_ref:
            payload['git_ref'] = git_ref
        if source_path:
            payload['source_path'] = source_path
        if job_env_vars:
            payload['job_env_vars'] = job_env_vars
        if timeout_seconds:
            payload['timeout_seconds'] = timeout_seconds
        if priority:
            payload['priority'] = priority

        response = self.session.post(
            f'{self.base_url}/api/v1/jobs',
            json=payload
        )
        response.raise_for_status()
        return response.json()

    def get_job(self, job_id: str) -> Dict[str, Any]:
        """Get a job by ID."""
        response = self.session.get(f'{self.base_url}/api/v1/jobs/{job_id}')
        response.raise_for_status()
        return response.json()

    def list_jobs(self, limit: int = 100, offset: int = 0) -> List[Dict[str, Any]]:
        """List all jobs."""
        response = self.session.get(
            f'{self.base_url}/api/v1/jobs',
            params={'limit': limit, 'offset': offset}
        )
        response.raise_for_status()
        result = response.json()
        return result.get('jobs', [])

    def cancel_job(self, job_id: str) -> None:
        """Cancel a running job."""
        response = self.session.put(f'{self.base_url}/api/v1/jobs/{job_id}/cancel')
        response.raise_for_status()

    def delete_job(self, job_id: str) -> None:
        """Delete a job."""
        response = self.session.delete(f'{self.base_url}/api/v1/jobs/{job_id}')
        response.raise_for_status()

    def wait_for_job(self, job_id: str, timeout: int = 300, poll_interval: int = 2) -> Dict[str, Any]:
        """
        Wait for a job to complete.

        Args:
            job_id: The job ID to wait for
            timeout: Maximum time to wait in seconds
            poll_interval: Time between status checks in seconds

        Returns:
            The final job object

        Raises:
            TimeoutError: If the job doesn't complete within the timeout
        """
        start_time = time.time()
        terminal_states = {'completed', 'failed', 'cancelled'}

        while time.time() - start_time < timeout:
            job = self.get_job(job_id)
            if job['status'] in terminal_states:
                return job
            time.sleep(poll_interval)

        raise TimeoutError(f"Job {job_id} did not complete within {timeout} seconds")


def example_git_job(client: ReactorcideClient):
    """Example: Create and monitor a job that clones a git repository"""
    print("\n=== Git Repository Job Example ===")

    job = client.create_job(
        name="Git Clone and Build",
        description="Clone a repository and run build commands",
        source_type="git",
        git_url="https://github.com/example/project.git",
        git_ref="main",
        job_command="make build && make test",
        runner_image="golang:1.21",
        job_env_vars={
            "GO111MODULE": "on",
            "CGO_ENABLED": "0"
        },
        queue_name="build-queue",
        timeout_seconds=600,
        priority=5
    )

    print(f"Created job: {job['job_id']}")
    print(f"Status: {job['status']}")
    print(f"Job submitted to Corndogs queue: {job.get('queue_name', 'default')}")

    # Monitor job progress
    print("\nMonitoring job progress...")
    try:
        completed_job = client.wait_for_job(job['job_id'], timeout=120)
        print(f"Job completed with status: {completed_job['status']}")

        if completed_job.get('logs_object_key'):
            print(f"Logs available at: {completed_job['logs_object_key']}")
        if completed_job.get('artifacts_object_key'):
            print(f"Artifacts available at: {completed_job['artifacts_object_key']}")

    except TimeoutError as e:
        print(f"Job monitoring timed out: {e}")
        # Optionally cancel the job
        client.cancel_job(job['job_id'])
        print("Job cancelled due to timeout")


def example_simple_job(client: ReactorcideClient):
    """Example: Create a simple echo job"""
    print("\n=== Simple Echo Job Example ===")

    job = client.create_job(
        name="Echo Test",
        description="A simple test job",
        source_type="copy",
        job_command="echo 'Hello from Reactorcide!' && ls -la",
        runner_image="alpine:latest",
        job_env_vars={
            "TEST_VAR": "test_value"
        }
    )

    print(f"Created job: {job['job_id']}")
    print(f"Status: {job['status']}")
    print("Job has been submitted to Corndogs for processing")

    return job


def example_list_jobs(client: ReactorcideClient):
    """Example: List all jobs"""
    print("\n=== Listing Jobs ===")

    jobs = client.list_jobs(limit=10)

    if not jobs:
        print("No jobs found")
        return

    for job in jobs:
        print(f"- {job['job_id']}: {job['name']}")
        print(f"  Status: {job['status']}")
        print(f"  Created: {job['created_at']}")
        print(f"  Queue: {job.get('queue_name', 'default')}")
        print()


def main():
    # Get configuration from environment variables
    api_url = os.getenv('REACTORCIDE_API_URL', 'http://localhost:8080')
    api_token = os.getenv('REACTORCIDE_API_TOKEN')

    if not api_token:
        print("Error: REACTORCIDE_API_TOKEN environment variable not set")
        print("\nTo create a test token:")
        print("1. Check coordinator_api/test/jobs_test.go for examples")
        print("2. Use the pattern: token_value = 'test-api-token-{user_id}'")
        print("3. Store SHA256 hash of token_value in database")
        print("4. Use original token_value as REACTORCIDE_API_TOKEN")
        sys.exit(1)

    # Create client
    client = ReactorcideClient(api_url, api_token)

    try:
        # Check API health
        print("Checking API health...")
        health = client.check_health()
        print(f"API Status: {health['status']}")
        print(f"Verification: {health.get('verification', {})}")

        # Run examples
        simple_job = example_simple_job(client)
        example_git_job(client)
        example_list_jobs(client)

        # Clean up - delete the simple job
        print(f"\nCleaning up - deleting job {simple_job['job_id']}...")
        client.delete_job(simple_job['job_id'])
        print("Job deleted successfully")

    except requests.HTTPError as e:
        print(f"HTTP Error: {e}")
        print(f"Response: {e.response.text}")
    except Exception as e:
        print(f"Error: {e}")


if __name__ == '__main__':
    main()